File: app/src/main/java/com/botijasoftware/utils/renderer/RenderQueue.java
package com.botijasoftware.utils.renderer;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;

import com.botijasoftware.utils.Camera;
import com.botijasoftware.utils.materials.Material;


public class RenderQueue {
	HashMap< Material, ArrayList<RenderItem>> queue = new HashMap< Material, ArrayList<RenderItem>> ();
	
	public void sort(Camera camera) {
	
	}
	
	public void add(RenderItem item) {
		
		if (queue.containsKey(item.material)) {
			queue.get(item.material).add(item);
		}
		else {
			ArrayList<RenderItem> list = new ArrayList<RenderItem>();
			list.add(item);
			queue.put(item.material, list);
		}
		
	}
	
	public void clear() {
		Collection<ArrayList<RenderItem>> set = queue.values();

		for (ArrayList<RenderItem> list : set) {
			list.clear();
		}
	    
	    queue.clear();
	}
	
	public int size() {
		return queue.size();
	}
	
	public ArrayList<Material> getMaterials() {
		Collection<Material> set = queue.keySet();
	    Iterator<Material> i = set.iterator();
	    
	    ArrayList<Material> list = new ArrayList<Material>();
	    		
	    while(i.hasNext()){
	    	Material m = i.next();
	    	list.add(m);
	    }
	    return list;
	}
	
	public ArrayList<RenderItem> getRenderItem(Material material) {
		if (queue.containsKey(material))
			return queue.get(material);
		else
			return new ArrayList<RenderItem>();
	}
}

File: app/src/main/java/com/botijasoftware/utils/renderer/RenderQueue.kt
package com.botijasoftware.utils.renderer

import java.util.ArrayList
import java.util.HashMap

import com.botijasoftware.utils.Camera
import com.botijasoftware.utils.materials.Material


class RenderQueue {
    internal var queue = HashMap<Material, ArrayList<RenderItem>>()

    fun sort(camera: Camera) {

    }

    fun add(item: RenderItem) {

        if (queue.containsKey(item.material)) {
            queue[item.material]!!.add(item)
        } else {
            val list = ArrayList<RenderItem>()
            list.add(item)
            queue.put(item.material, list)
        }

    }

    fun clear() {
        val set = queue.values

        for (list in set) {
            list.clear()
        }

        queue.clear()
    }

    fun size(): Int {
        return queue.size
    }

    val materials: ArrayList<Material>
        get() {
            val set = queue.keys
            val i = set.iterator()

            val list = ArrayList<Material>()

            while (i.hasNext()) {
                val m = i.next()
                list.add(m)
            }
            return list
        }

    fun getRenderItem(material: Material): ArrayList<RenderItem>? {
        if (queue.containsKey(material))
            return queue[material]
        else
            return ArrayList()
    }
}